include/zephyr/net/dsa_core.h

```diff
@@ -0,0 +1,272 @@
+	COND_CODE_1(DT_NUM_PINCTRL_STATES(port),                                                   \
+				(PINCTRL_DT_DEFINE(port);), (EMPTY))                               \
+	const struct dsa_port_config dsa_##dsa##_port_##port##_config = {                          \
+		.mac_addr = DT_PROP_OR(port, local_mac_address, {0}),                              \
+		.port_idx = DT_REG_ADDR(port),                                                     \
+		.phy_dev = (COND_CODE_1(DT_NODE_HAS_PROP(port, phy_handle),                        \
+					DEVICE_DT_GET(DT_PHANDLE(port, phy_handle)),               \
+					NULL)),                                                    \
+		.phy_mode = DT_PROP_OR(port, phy_connection_type, ""),                             \
+		.ethernet_connection = (COND_CODE_1(DT_NODE_HAS_PROP(port, ethernet),              \
+					(DEVICE_DT_GET(DT_PHANDLE(port, ethernet))), NULL)),       \
+		.pincfg = COND_CODE_1(DT_NUM_PINCTRL_STATES(port),                                 \
+				(PINCTRL_DT_DEV_CONFIG_GET(port)), NULL),                          \
+	};                                                                                         \
+	NET_DEVICE_INIT_INSTANCE(                                                                  \
+		CONCAT(dsa_, dsa, _port_, port),                                                   \
+		"swp" STRINGIFY(DT_REG_ADDR(port)), DT_REG_ADDR(port), dsa_port_initialize, NULL,  \
+				&dsa_context_##dsa, &dsa_##dsa##_port_##port##_config,             \
+				CONFIG_ETH_INIT_PRIORITY, &dsa_eth_api, ETHERNET_L2,               \
+				NET_L2_GET_CTX_TYPE(ETHERNET_L2), NET_ETH_MTU);
+
+#define DSA_INIT_INSTANCE(n, _dapi, data)                                                          \
+	struct dsa_context dsa_context_##n = {                                                     \
+		.dapi = _dapi,                                                                     \
+		.prv_data = (void *)data,                                                          \
+		.init_ports = 0,                                                                   \
+		.num_ports = DT_INST_CHILD_NUM_STATUS_OKAY(n),                                     \
+	};                                                                                         \
+	DT_INST_FOREACH_CHILD_STATUS_OKAY_VARGS(n, DSA_PORT_INIT_INSTANCE, n);
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/** DSA context data */
+struct dsa_context {
+	/** Pointers to all DSA user network interfaces */
+	struct net_if *iface_user[DSA_PORT_MAX_COUNT];
+
+	/** Pointer to DSA conduit network interface */
+	struct net_if *iface_conduit;
+
+	/** DSA specific API callbacks - filled in the switch IC driver */
+	struct dsa_api *dapi;
+
+	/** Number of ports in the DSA switch */
+	uint8_t num_ports;
+
+	/** Number of initialized ports in the DSA switch */
+	uint8_t init_ports;
+
+	/** Instance specific data */
+	void *prv_data;
```

Comment:
Minor nit, but could we change the ordering of these fields to avoid hole in the struct because of alignment. This would not affect the size of the struct atm, but could if new fields are added to it. So suggesting
```
	/** Instance specific data */
	void *prv_data;

	/** Number of ports in the DSA switch */
	uint8_t num_ports;

	/** Number of initialized ports in the DSA switch */
	uint8_t init_ports;
```